### PR TITLE
[Silabs] [WiFi] Fixing the retry logic for the WiFi Devices

### PR DESCRIPTION
**Pull request by chirag-silabs**
The retry logic was broken for the wifi devices, the following consist of following changes:

- The retry was only working for 5 times on power cycle
- moving to timer based instead of the vtaskDelay
- Fixes all the NCP and SoC Wifi combos
- Moving the retry logic from the src/platform/silabs to examples/platform/silabs

Tested on NCP and SoC combs

### CODE REVIEW

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 1]
```diff
@@ -166,57 +171,46 @@ void wfx_ip_changed_notify(int got_ip)
...
+            ChipLogProgress(DeviceLayer, "wfx_retry_connection : Next attempt after %d Seconds",
                             CONVERT_MS_TO_SEC(WLAN_RETRY_TIMER_MS));
-#if SL_ICD_ENABLED
-            // TODO: cleanup the retry logic MATTER-1921
-            if (!chip::Server::GetInstance().GetCommissioningWindowManager().IsCommissioningWindowOpen())
+            if (osTimerStart(sRetryTimer, pdMS_TO_TICKS(WLAN_RETRY_TIMER_MS)) != osOK)
             {
-                wfx_rsi_power_save(RSI_SLEEP_MODE_8, STANDBY_POWER_SAVE_WITH_RAM_RETENTION);
+                ChipLogProgress(DeviceLayer, "Failed to start retry timer");
+                return;
```
Comment (mkardous-silabs) [inline examples/platform/silabs/wifi/wfx_notify.cpp:203]: What happens if we fail to start the timer? we never retry?

File: examples/platform/silabs/efr32/wf200/host_if.cpp [hunk 2]
```diff
@@ -541,13 +534,8 @@ static void wfx_events_task(void * p_arg)
...
         if (flags & SL_WFX_RETRY_CONNECT)
         {
-            if (!retryInProgress)
-            {
-                retryInProgress = true;
-                wfx_retry_interval_handler(is_wifi_disconnection_event, retryJoin);
-                SILABS_LOG("WFX sending the connect command");
-                wfx_connect_to_ap();
-            }
+            SILABS_LOG("WFX sending the connect command");
```
Comment (mkardous-silabs) [inline examples/platform/silabs/efr32/wf200/host_if.cpp:537]:
```suggestion
```

File: examples/platform/silabs/efr32/rs911x/rsi_if.c [hunk 3]
```diff
@@ -586,27 +573,16 @@ static void wfx_rsi_do_join(void)
...
-                           wfx_rsi.join_retries);
-                break; // exit while loop
-            }
+            wfx_rsi.dev_state &= ~WFX_RSI_ST_STA_CONNECTING;
+            SILABS_LOG("%s: rsi_wlan_connect_async failed with status: %02x on try %d", __func__, status, wfx_rsi.join_retries);
+            wfx_retry_connection(++wfx_rsi.join_retries);
+        }
+        else
+        {
+            SILABS_LOG("%s: starting JOIN to %s after %d tries\n", __func__, (char *) &wfx_rsi.sec.ssid[0], wfx_rsi.join_retries);
```
Comment (mkardous-silabs) [inline examples/platform/silabs/efr32/rs911x/rsi_if.c:585]: Avoid using SILABS_LOGS.

[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/efr32/rs911x/rsi_if.c]: done

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 4]
```diff
@@ -67,6 +67,15 @@ void wfx_started_notify()
     sl_wfx_startup_ind_t evt;
     sl_wfx_mac_address_t mac;
 
+    // Creating a timer which will be used to retry connection with AP
+    sRetryTimer = osTimerNew(RetryConnectionTimerHandler, osTimerOnce, NULL, NULL);
+    if (sRetryTimer == NULL)
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp]: done

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 5]
```diff
@@ -29,32 +32,29 @@
...
-#include <app/server/Server.h>
 
 using namespace ::chip;
 using namespace ::chip::DeviceLayer;
 
-#include <lib/support/logging/CHIPLogging.h>
-
-extern uint32_t retryInterval;
+static uint16_t retryInterval = WLAN_MIN_RETRY_TIMER_MS;
+static osTimerId_t sRetryTimer;
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp]: done

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 6]
```diff
@@ -29,32 +32,29 @@
...
 /*
  * Notifications to the upper-layer
  * All done in the context of the RSI/WiFi task (rsi_if.c)
  */
 
+static void RetryConnectionTimerHandler(void * arg)
+{
+#if CHIP_CONFIG_ENABLE_ICD_SERVER && SLI_SI91X_MCU_INTERFACE
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp:51]: added

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 7]
```diff
@@ -29,32 +32,29 @@
...
+    // TODO: remove this once it is fixed in the wifi sdk SI91X-15845
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp]: Removed the comment

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 8]
```diff
@@ -90,13 +99,15 @@ void wfx_connected_notify(int32_t status, sl_wfx_mac_address_t * ap)
 {
     sl_wfx_connect_ind_t evt;
 
+    ChipLogProgress(DeviceLayer, "wfx_connected_notify : started.");
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp]: done

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 9]
```diff
@@ -90,13 +99,15 @@ void wfx_connected_notify(int32_t status, sl_wfx_mac_address_t * ap)
...
+
     if (status != SUCCESS_STATUS)
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp:98]: done

File: examples/platform/silabs/wifi/wfx_notify.cpp [hunk 10]
```diff
@@ -29,32 +32,29 @@
...
-#include <app/server/Dnssd.h>
-#include <app/server/Server.h>
 
 using namespace ::chip;
 using namespace ::chip::DeviceLayer;
 
-#include <lib/support/logging/CHIPLogging.h>
-
-extern uint32_t retryInterval;
+static uint16_t retryInterval = WLAN_MIN_RETRY_TIMER_MS;
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/wifi/wfx_notify.cpp]: done

File: examples/platform/silabs/efr32/wf200/host_if.cpp (hunk 2 above)
Comment (chirag-silabs) [reply examples/platform/silabs/efr32/wf200/host_if.cpp:537]: @mkardous-silabs  all the logging will be fixed as a separate PR which is currently being worked on

File: examples/platform/silabs/efr32/rs911x/rsi_if.c (hunk 3 above)
Comment (chirag-silabs) [reply examples/platform/silabs/efr32/rs911x/rsi_if.c]: @mkardous-silabs all the logging will be fixed as a separate PR which is currently being worked on

File: examples/platform/silabs/efr32/rs911x/rsi_if.c [hunk 11]
```diff
@@ -586,27 +573,16 @@ static void wfx_rsi_do_join(void)
...
-            {
-                SILABS_LOG("%s: starting JOIN to %s after %d tries\n", __func__, (char *) &wfx_rsi.sec.ssid[0],
-                           wfx_rsi.join_retries);
-                break; // exit while loop
-            }
+            wfx_rsi.dev_state &= ~WFX_RSI_ST_STA_CONNECTING;
+            SILABS_LOG("%s: rsi_wlan_connect_async failed with status: %02x on try %d", __func__, status, wfx_rsi.join_retries);
+            wfx_retry_connection(++wfx_rsi.join_retries);
+        }
+        else
```
[the remark this answers is not in the document]
Comment (chirag-silabs) [reply examples/platform/silabs/efr32/rs911x/rsi_if.c]: done

**Review by jmartinez-silabs — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Pull request merged**